tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -11117,40 +11117,44 @@ void hlsl::DiagnoseRegisterType(clang::Sema *self, clang::SourceLocation loc,
 }
 
 // Check HLSL member call constraints
-bool Sema::DiagnoseHLSLMethodCall(const CXXMethodDecl *MD, SourceLocation Loc) {
+bool Sema::DiagnoseHLSLMethodCall(const CXXMethodDecl *MD, SourceLocation Loc,
+                                  bool SkipUnused) {
   if (MD->hasAttr<HLSLIntrinsicAttr>()) {
-    // If this is a call to FinishedCrossGroupSharing then the Input record
-    // must have the NodeTrackRWInputSharing attribute
     hlsl::IntrinsicOp opCode =
         (IntrinsicOp)MD->getAttr<HLSLIntrinsicAttr>()->getOpcode();
-    if (opCode == hlsl::IntrinsicOp::MOP_FinishedCrossGroupSharing) {
-      const CXXRecordDecl *NodeRecDecl = MD->getParent();
-      // Node I/O records are templateTypes
-      const ClassTemplateSpecializationDecl *templateDecl =
-          cast<ClassTemplateSpecializationDecl>(NodeRecDecl);
-      auto &TemplateArgs = templateDecl->getTemplateArgs();
-      DXASSERT(TemplateArgs.size() == 1,
-               "Input record types need to have one template argument");
-      auto &Rec = TemplateArgs.get(0);
-      clang::QualType RecType = Rec.getAsType();
-      RecordDecl *RD = RecType->getAs<RecordType>()->getDecl();
-      if (!RD->hasAttr<HLSLNodeTrackRWInputSharingAttr>()) {
-        Diags.Report(Loc, diag::err_hlsl_wg_nodetrackrwinputsharing_missing);
-        return true;
+    if (SkipUnused) {
```

Comment:
It seems like we wouldn't want code executing from both cases, so why can't we make this case a separate function?  That way arguments can be added as needed to handle the call graph traversal case without impacting parse-time diagnostics.

Reply:
Fixed.